### PR TITLE
Move flambda-backend specific flags out of ocaml/ subdirectory

### PR DESCRIPTION
**Pull request by gretay-js**
The goal is to remove flambda-backend specific things from `ocaml/` subdirectory to make it easier to track upstream. This PR handles flambda-backend specific flags in command line and OCAMLPARAM.  The latter is handled by extending the interface of `Compenv` with the ability to set a custom param parser.

Adds `Flambda_backend_args` `Flambda_backend_flags` in `driver` subdirectory to mimic `Main_args` and `Clflags` from `ocaml`. `Flambda_backend_args` can probably be simplified along the lines of `Arch.command_line_options`, but it seems useful to keep the ability to override the flags, as is done in `Opttopmain` in this PR.

The new modules are currently only used in `ocamloptcomp`  library. It means that the flags are supported in `ocamlopt` only, not in `ocamlc` or any other bytecode-based tools. An unfortunate consequence of it is that if we have, say, `ocamlcfg=1` in OCAMLPARAM, and run `ocamlc`, it'll spew warnings like this for every file:
```
File "_none_", line 1:    
Warning 46 [bad-env-variable]: illegal environment variable OCAMLPARAM : Warning: discarding value of variable "ocamlcfg" in OCAMLPARAM
```
To silence these warnings, we can add the flag `can_discard=ocamlcfg` to OCAMLPARAM globally by writing it to a "config" file named `ocaml_compiler_internal_params` in `lib/ocaml` subdirectory of the compiler's installation. To build the compiler itself with an flambda-backend flag in OCAMLPARAM, the config file also needs to be installed for each of the build stages. There is a workaround for stage0, because the compiler looks for this file only in the `install` directory. Alternative solutions we have considered are disabling warning 46 globally except for ocamlopt (but this PR is safer) and modifying bytecode tools to install the new params, which seems too invasive.

The first commit only moves `-ocamlcfg` flag. If this approach is fine, I'll move the others.

### CODE REVIEW

File: tools/ocaml_compiler_internal_params [hunk 1]
```diff
@@ -0,0 +1,27 @@
+* : can-discard = ocamlcfg
```
Comment (mshinwell) [inline tools/ocaml_compiler_internal_params:1]: The `tools` directory doesn't seem like the right place for this file.  Maybe it should just live at the root?

File: ocaml/stdlib/dune [hunk 2]
```diff
@@ -186,6 +186,7 @@
     weak.ml
     weak.mli
 
+    ocaml_compiler_internal_params
```
Comment (mshinwell) [inline ocaml/stdlib/dune:189]: I don't think this should be in the `stdlib` dune file -- it isn't related to the stdlib.

File: ocaml/utils/clflags.ml [hunk 3]
```diff
@@ -52,7 +52,6 @@ and debug = ref false                   (* -g *)
 and debug_full = ref false              (* For full DWARF support *)
 and unsafe = ref false                  (* -unsafe *)
 and use_linscan = ref false             (* -linscan *)
-and use_ocamlcfg = ref false            (* -ocamlcfg *)
 and link_everything = ref false         (* -linkall *)
```
Comment (mshinwell) [inline ocaml/utils/clflags.ml:55]: I think it would be worth doing a check (if you haven't already) on the final version of this PR by diffing this file against the 4.12 upstream version, just to make sure nothing has been missed.

File: ocaml/.gitignore [hunk 4]
```diff
@@ -201,6 +201,7 @@ _build
 /stdlib/labelled-*
 /stdlib/caml
 /stdlib/sys.ml
+/stdlib/ocaml_compiler_internal_params
```
Comment (mshinwell) [inline ocaml/.gitignore:204]: I don't understand why this has to be gitignored.

File: driver/flambda_backend_args.mli [hunk 5]
```diff
@@ -0,0 +1,97 @@
+(**************************************************************************)
```
Comment (mshinwell) [inline driver/flambda_backend_args.mli:1]: This is clearly the wrong copyright notice -- same for `flambda_backend_flags.ml`.  We will harmonise these in due course but for the moment please take one from the Flambda 2 code.

File: driver/flambda_backend_args.ml [hunk 6]
```diff
@@ -0,0 +1,757 @@
...
+(*                                                                        *)
+(*   Copyright 1998 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+(* Keep in sync with tools/ocaml_compiler_internal_params. *)
```
Comment (mshinwell) [inline driver/flambda_backend_args.ml:15]: These don't appear to be in sync at the moment?
Also please document the exact rule about what needs to go in the file -- presumably it's only the ones that are parsed in `OCAMLPARAM`.  (Some are not e.g. `-drawfexpr`.)

Comment (mshinwell) [reply driver/flambda_backend_args.ml:15]: oh actually maybe they are in sync and I was misreading.  Can we generate the file somehow so we don't have this in-sync requirement?

File: Makefile.in [hunk 7]
```diff
@@ -228,6 +236,10 @@ config_files: ocaml-stage1-config.status
 	/bin/echo -n "( `grep \"^SHAREDLIB_CFLAGS=\" ocaml/Makefile.config \
 		| sed 's/^SHAREDLIB_CFLAGS=//'` )" > sharedlib_cflags.sexp
 
+.PHONY: ocaml_compiler_internal_param
```
Comment (mshinwell) [inline Makefile.in:239]: Does this target name deliberately lack the final "s"?  I'd prefer it to have the "s" but maybe something will go wrong.

File: Makefile.in [hunk 8]
```diff
@@ -31,15 +31,23 @@ dune=@dune@
...
 .PHONY: stage0
-stage0: _build0/config.status
+stage0: _build0/config.status ocaml_compiler_internal_param
 	rm -f .rsync-output
 	rsync -i -a --filter=':- $$(pwd)/ocaml/.gitignore' \
 	  $$(pwd)/ocaml/ $$(pwd)/_build0 \
 	  | grep -v '/$$' \
 	  | tee .rsync-output
-	if [ -s .rsync-output ] || ! [ -d @stage0_prefix@ ]; then \
+	if [ -s .rsync-output ] || ! [ -d @stage0_prefix@/bin ]; then \
```
Comment (mshinwell) [inline Makefile.in:47]: Why is this change needed?

File: Makefile.in [hunk 9]
```diff
@@ -228,6 +236,10 @@ config_files: ocaml-stage1-config.status
 	/bin/echo -n "( `grep \"^SHAREDLIB_CFLAGS=\" ocaml/Makefile.config \
 		| sed 's/^SHAREDLIB_CFLAGS=//'` )" > sharedlib_cflags.sexp
 
+.PHONY: ocaml_compiler_internal_param
+ocaml_compiler_internal_param: tools/ocaml_compiler_internal_params
+	cp tools/ocaml_compiler_internal_params ocaml/stdlib/
```
Comment (mshinwell) [inline Makefile.in:241]: I now see why the gitignore change below is needed, but I'd really like to avoid modifying the `ocaml/` directory during the build if possible.  (@lukemaurer has already done some work towards this.)  Can this file be found anywhere else?

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Comment by mshinwell** [comment]
A fair amount of complexity has been introduced here compared to the approach of disabling the warning for everything except `ocamlopt`.  Could we maybe get this change in with the warning disabled in that manner, and then a separate PR can be made with the flags file changes?

File: ocaml/stdlib/dune (hunk 2 above)
Comment (gretay-js) [reply ocaml/stdlib/dune:189]: That's where the compiler expects it to be installed.

File: Makefile.in (hunk 8 above)
Comment (gretay-js) [reply Makefile.in:47]: yeah, I think so.. the action here creates @stage_prefix@/lib so it can copy ocaml_compiler_internal_params into it, even if install didn't run.

File: Makefile.in (hunk 9 above)
Comment (gretay-js) [reply Makefile.in:241]: I can rewrote the Makefile to copy this file to the `@stageX_prefix@/lib directly, without ever going through src or build, would it be better?

**Comment by gretay-js** [comment]
> A fair amount of complexity has been introduced here compared to the approach of disabling the warning for everything except `ocamlopt`. Could we maybe get this change in with the warning disabled in that manner, and then a separate PR can be made with the flags file changes?

Right, I'll split this PR.

File: ocaml/utils/clflags.ml (hunk 3 above)
Comment (gretay-js) [reply ocaml/utils/clflags.ml:55]: If you mean in comparison to ocaml/ subdirectory, I checked when I've move the module Clflags.Flambda2 and from compenv. Now if anyone adds any flags on `main` branch, this PR will be in conflict.

File: driver/flambda_backend_args.mli (hunk 5 above)
Comment (gretay-js) [reply driver/flambda_backend_args.mli:1]: fixed.

File: driver/flambda_backend_args.mli [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  module Opttopmain: Opttop_options
+end
+
+module Extra_params : sig
+  (* This is installed unconditionally for native and bytecode and
+     all tools that consume OCAMLPARAM, to avoid warnings about flags
+     that are only supported in flambda-backend,
+     similarly to the treatment of native-only flags in
+     [ocaml/driver/compenv.ml] *)
+  (** [read_param ppf pos name value] returns whether the param was handled.  *)
```
Comment (mshinwell) [inline driver/flambda_backend_args.mli:95]: Is this comment still accurate?  (I'm not sure it buys us much actually, maybe it could be removed.  The code says what's going on.)

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: driver/flambda_backend_args.mli (hunk 10 above)
Comment (gretay-js) [reply driver/flambda_backend_args.mli:95]: The first comment (L90-94) is stale, I'll remove it in a moment, thanks!
The second comment is accurate. If this function is used by setting `Compenv.extra_params`, it'll be called by `Compenv.read_one_param` **before** any other flags are match, which allows us to override existing options such as oclassic and o3.

**Review by xclerc — APPROVED** [review]

**Comment by mshinwell** [comment]
I've resolved conversations from my review on the part of the original PR that is being deferred (i.e. using the separate params file).

**Review by mshinwell — APPROVED** [review]
